ferretdb/ferretdb.go

```diff
@@ -0,0 +1,149 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ferretdb
+
+import (
+	"context"
+	"net"
+	"net/url"
+	"time"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+
+	"github.com/FerretDB/FerretDB/internal/clientconn"
+	"github.com/FerretDB/FerretDB/internal/handlers/common/register"
+	"github.com/FerretDB/FerretDB/internal/util/debug"
+	"github.com/FerretDB/FerretDB/internal/util/logging"
+	"github.com/FerretDB/FerretDB/internal/util/must"
+	"github.com/FerretDB/FerretDB/internal/util/version"
+)
+
+// Config contains a PostgreSQL backend connection string.
+type Config struct {
+	PostgreSQLConnectionString string
+}
+
+// FerretDB proxy.
+type FerretDB struct {
+	config   Config
+	mongoURL string
+}
+
+// New returns a new FerretDB.
+func New(conf Config) FerretDB {
+	return FerretDB{
+		config:   conf,
+		mongoURL: transformConnStringPgToMongo(conf.PostgreSQLConnectionString),
+	}
+}
+
+// Run runs the FerretDB proxy as a library with
+// * error level logging
+// * monitoring disabled
+// * handler PostgreSQL.
+func (fdb *FerretDB) Run(ctx context.Context, conf Config) (string, error) {
+	listenAddr := "127.0.0.1:27017"
+	proxyAddr := "127.0.0.1:37017"
+	debugAddr := "127.0.0.1:8088"
+	mode := clientconn.NormalMode
+	handler := "pg"
+	testConnTimeout := time.Duration(0)
+
+	_, ok := register.HandlerFunc["pg"]
+	if !ok {
+		panic("no pg handler registered")
+	}
+
+	logging.Setup(zapcore.ErrorLevel)
+	logger := zap.L()
+
+	info := version.Get()
+
+	startFields := []zap.Field{
+		zap.String("version", info.Version),
+		zap.String("commit", info.Commit),
+		zap.String("branch", info.Branch),
+		zap.Bool("dirty", info.Dirty),
+	}
+	for _, k := range info.BuildEnvironment.Keys() {
+		v := must.NotFail(info.BuildEnvironment.Get(k))
+		startFields = append(startFields, zap.Any(k, v))
+	}
+
+	go debug.RunHandler(ctx, debugAddr, logger.Named("debug"))
+
+	newHandler := register.HandlerFunc[handler]
+	if newHandler == nil {
+		logger.Sugar().Fatalf("Unknown backend handler %q.", handler)
+	}
+	h, err := newHandler(&register.NewHandlerOpts{
+		PostgreSQLConnectionString: conf.PostgreSQLConnectionString,
+		Ctx:                        ctx,
+		Logger:                     logger,
+	})
+	if err != nil {
+		logger.Fatal(err.Error())
+	}
+	defer h.Close()
+
+	l := clientconn.NewListener(&clientconn.NewListenerOpts{
+		ListenAddr:      listenAddr,
+		ProxyAddr:       proxyAddr,
+		Mode:            clientconn.Mode(mode),
+		Handler:         h,
+		Logger:          logger,
+		TestConnTimeout: testConnTimeout,
+	})
+
+	err = l.Run(ctx)
+	if err == nil || err == context.Canceled {
+		logger.Info("Listener stopped")
+	} else {
+		logger.Error("Listener stopped", zap.Error(err))
+	}
+	return fdb.mongoURL, nil
```

Comment:
So this function returns connection string when listener already stopped. That could not possibly work

Reply:
> So this function returns connection string when listener already stopped. That could not possibly work

right you are, I duly followed your recommendation on returning the connection string from the `Run` function. I'll get back what was before - a separate function one.